add onRetryCallback callback function

shamil opened the pull request:
Implements the suggested solutoin in https://github.com/grpc-ecosystem/go-grpc-middleware/issues/356

johanbrandhorst commented on interceptors/retry/options.go:
```diff
@@ -75,6 +87,22 @@ func WithBackoffContext(bf BackoffFuncContext) CallOption {
 	}}
 }
 
+// WithOnRetryCallback sets the `OnRetryCallback` used to add additional logic when retry occurs.
+func WithOnRetryCallback(fn OnRetryCallback) CallOption {
+	return CallOption{applyFunc: func(o *options) {
+		o.onRetryCallback = OnRetryCallbackContext(func(ctx context.Context, attempt uint, err error) {
+			fn(attempt, err)
+		})
+	}}
+}
+
+// WithOnRetryCallbackContext sets the `OnRetryCallbackContext` used to add additional logic when retry occurs.
+func WithOnRetryCallbackContext(fn OnRetryCallbackContext) CallOption {
+	return CallOption{applyFunc: func(o *options) {
+		o.onRetryCallback = fn
+	}}
+}
```
Re: merge these into one

johanbrandhorst commented on interceptors/retry/retry.go:
```diff
@@ -48,6 +48,7 @@ func UnaryClientInterceptor(optFuncs ...CallOption) grpc.UnaryClientInterceptor
 				return nil
 			}
+			callOpts.onRetryCallback(callCtx, attempt, lastErr)
 			logTrace(parentCtx, "grpc_retry attempt: %d, got err: %v", attempt, lastErr)
```
Maybe we can make this log here the default retry callback behaviour while we're here?

shamil replied:
I can do it, but If we do this, we will lose this trace log once the custom callback will be set bu user.

johanbrandhorst replied:
We'll document that as the default behaviour so that the user knows that it will disappear if they override the callback. They can always manually add it back again.

shamil replied:
I can also put this in the `WithOnRetryCallback` before invoking the user's logic, if this makes sense.

johanbrandhorst replied:
No I think one of the things I want to do here is allow the user to disable this log behaviour too.

shamil replied:
All right. Done

johanbrandhorst commented on interceptors/retry/options.go:
```diff
@@ -45,6 +48,15 @@ type BackoffFunc func(attempt uint) time.Duration
+//
+type OnRetryCallback func(attempt uint, err error)
+
+// OnRetryCallback called in retry attempts flow, which can be used to add additiional logic when retry occurs
+//
+// The context can be used to extract request scoped metadata and context values.
```
After merging these, lets simplify this comment a bit too.

```suggestion
// OnRetryCallback is the type of function called when a retry occurs.
```

johanbrandhorst commented on interceptors/retry/options.go:
```diff
@@ -75,6 +87,22 @@ func WithBackoffContext(bf BackoffFuncContext) CallOption {
 	}}
 }
 
+// WithOnRetryCallback sets the `OnRetryCallback` used to add additional logic when retry occurs.
```
```suggestion
// WithOnRetryCallback sets the callback to use when a retry occurs.
//
// By default, no action is performed on retry.
```

johanbrandhorst reviewed (changes requested):
Thanks for taking this on! A few suggestions.

shamil commented:
Thanks for review @johanbrandhorst
Addressed all the comments, had only one concern about making the trace log the default callback behaviour. Let me know how you want to proceed and I will make the require changes.

johanbrandhorst reviewed (approved):
LGTM

johanbrandhorst commented:
Thanks for your contribution!
